packages/libs/deer/src/error.rs

```diff
@@ -1,15 +1,196 @@
+//! # Error Structure
+//!
+//! Every error emitted by `deer` (and potentially other tooling) follows the same schema when
+//! serialized:
+//!
+//! ```json5
+//! {
+//!     "ns": "deer", // namespace of the error
+//!     "id": ["type"], // unique id across the namespace
+//!     "properties": {} // object of machine readable properties related to id
+//!     "message": "" // human readable message
+//! }
+//! ```
+//!
+//! Optionally errors also emit a json-schema type, which can be used to present the user with all
+//! possible errors.
```

Comment:
Is the error restricted to JSON schema types or can this be anything implementing a certain trait?

Reply:
I am still unsure about this. I should move this into a TODO instead. This is likely going to be part of 0.2.

I have multiple ideas:
* new crate: `deer-schema`, which leverages traits and proc-macros for schemas, instead of only outputting `json-schema`, one can use *traits* and a rust DSL to define properties, which then can be compiled into different schema languages, like json-schema draft 4 or 2020-12, or OpenAPI 3.0 or whatever.
* existing crates like `valico` or `schemars` this has up and downsides, mostly `valico` is super heavy, and `schemars` API is not really exposed, hard to work with, and discouraged from implementation in libraries until 1.0 is reached
* untyped schema: just let the user decide. This is the least effort (`BTreeMap<String, Box<dyn erased_serde::Serialize>>`), but it can lead to incorrect schemas and does not support multiple schema types.

Reply:
Thanks for clarifying. Yes, I think a TODO would be good (does not need to be as long as your answer here)

Reply:
fixed in [29cc757](https://github.com/hashintel/hash/pull/1286/commits/29cc7571e8b74c4178e7a4bf50c2e1093d389c6b)